lib/seahorn/CexExeGenerator.cc

```diff
@@ -37,122 +37,130 @@ bool extractArrayContents(Expr e, kv &out, Expr &defaultValue) {
-  } else if (auto *sizeConst = dyn_cast<ConstantInt>(sizeArg)) {
-    expr::mpz_class sz = toMpz(sizeConst);
-    size = expr::mkTerm<expr::mpz_class>(sz, m_trace.engine().efac());
-  } else {
-    LOG("cex",
-        errs() << "unhandled Value of memhavoc size: " << *sizeArg << "\n");
-    return;
-  }
-  // get info of ptr to havoc
-  const Value *hPtrAlloc = cs.getArgOperand(0)->stripPointerCasts();
-  Expr hPtrStart;
-  if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtrAlloc)) {
-    hPtrStart = m_trace.eval(loc, *hPtrAllocInst, true);
-  } else {
-    LOG("cex",
-        errs() << "unhandled Value of memhavoc ptr: " << *hPtrAlloc << "\n");
-    return;
+class BmcTraceVisitor : public InstVisitor<BmcTraceVisitor<Trace>> {
+  CexExeGenerator<Trace> &m_cex;
+  unsigned m_loc;
+
+public:
+  BmcTraceVisitor(CexExeGenerator<Trace> &cex) : m_cex(cex) {}
+
+  void setLoc(unsigned loc) { m_loc = loc; }
+
+  void visitMemhavoc(CallSite CS) {
+    Instruction &I = *CS.getInstruction();
+    auto *CF = getCalledFunction(CS);
+    // previous instruction should be
+    // shadow.mem.load(i32 x, i32 %sm_n, i8* null)
+    if (I.getPrevNonDebugInstruction() == nullptr)
+      return;
+    const Instruction *prevI = I.getPrevNonDebugInstruction();
+    if (const CallInst *prevCI = dyn_cast<CallInst>(prevI)) {
+      ImmutableCallSite prevCS(prevCI);
+      const Function *prevF = prevCS.getCalledFunction();
+      if (!(prevF && prevF->getName().equals("shadow.mem.load"))) {
+        LOG("cex", ERR << "Skipping harness for memhavoc"
+                       << " because shadow.mem.load cannot be found\n");
+        return;
+      }
+      // get memhavoc content from corresponding shadow mem region
+      auto *shadowMemI = dyn_cast<Instruction>(prevCS.getArgOperand(1));
+      if (!shadowMemI)
+        return;
+      Expr shadowMem = m_cex.trace().eval(m_loc, *shadowMemI, true);
+      // get memhavoc size from second operand of memhavoc
+      const Value *sizeArg = CS.getArgOperand(1);
+      Expr size;
+      if (auto *sizeI = dyn_cast<Instruction>(sizeArg)) {
+        size = m_cex.trace().eval(m_loc, *sizeI, true);
+      } else if (auto *sizeConst = dyn_cast<ConstantInt>(sizeArg)) {
+        expr::mpz_class sz = toMpz(sizeConst);
+        size = expr::mkTerm<expr::mpz_class>(sz, m_cex.trace().engine().efac());
+      } else {
+        LOG("cex",
+            ERR << "unhandled Value of memhavoc size: " << *sizeArg << "\n");
+        return;
+      }
+      // get info of ptr to havoc
+      const Value *hPtrAlloc = CS.getArgOperand(0)->stripPointerCasts();
+      Expr hPtrStart;
+      if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtrAlloc)) {
+        hPtrStart = m_cex.trace().eval(m_loc, *hPtrAllocInst, true);
+      } else {
+        LOG("cex",
+            ERR << "unhandled Value of memhavoc ptr: " << *hPtrAlloc << "\n");
+        return;
+      }
+      Expr shadowMemRaw = m_cex.trace().engine().getPtrAddressable(shadowMem);
```

Comment:
This is wrong. 

`getAddressable` takes a pointer as input and returns an address. But `shadowMem` is not a pointer, it is a memory map, so, probably, a structure of 3 maps. Not sure why this works...

There should be a different function to get `raw` part of the memory map.

Reply:
the `Expr` encoding of a memory map looks very similar to pointer encoding, so I just used `getPtrAddressable`

Reply:
they both happen to be structs with 3 elements. But this is only in some configurations. The fields themselves are very different. In our default configuration, a pointer is `{base, offset, size}` and `getAddressable` returns `base + offset`. Not sure how this works for memory map which is `{raw, mem_offset, nem_size}` where `raw`, `mem_offset`, and `mem_size` are all memory maps themselves.

Reply:
@priyasiddharth what is preferred way to get raw memory from the manager?

Reply:
Assuming given an expression which is a `MemValTy` we want the raw part.
Currently there is no API for this.
Discussing with Xiang, best is to add a `ExprToRaw(e)` method to `OpSemMemManagerBase` and implement it only in the needed `MemManagers`

Reply:
Name should be something like `getRawMem` to be similar to `getAddressable` for ptr. It converts memory map to its raw part (i.e., main memory without shadows). The term `Expr` should not be in the name since `Expr` is just a low level data structure, not a concept at this level.

Reply:
added `getRawMem` for all memory managers in use, now `shadowMemRaw` is extracted with that instead; the function checks whether an `Expr` is in the right form of `MemValTy`, if so, return `getRaw` of the MemValTy. @priyasiddharth please take a look on the new mem manager functions.